### PR TITLE
[wgpu-hal] Inline RayQuery Support 

### PR DESCRIPTION
**Pull request by daniel-keitel**
**Checklist**

- [x] Run `cargo clippy`.
- [x] Run `RUSTFLAGS=--cfg=web_sys_unstable_apis cargo clippy --target wasm32-unknown-unknown` if applicable.
- [x] Add change to CHANGELOG.md. See simple instructions inside file.
- [x] Consider `unsafe_ops_in_unsafe_fn` #3044
- [x] Consider DX12 for API compatibility
- [x] Acceleration Structure Updates

**Future Work**
- [ ] Acceleration Structure Compaction
- [ ] DX 12 Implementation
- [ ] Ray Tracing Pipeline #3607

**Connections**
#1040
#3020 original PR
gfx-rs/naga#2256

**Description**
Basic Vulkan ray query and acceleration structure support.

This pull request is the continuation of @expenses PR #3020.

The Api is designed to work with a DirectX12 Implementation.

image

**Testing**
Tested with example application (ray-traced-triangle)

### CODE REVIEW

**Comment by expenses** [comment]
Just did a brief look over this again - I think the code is fine but the suggestions from the original PR should probably still be applied.

**Comment by daniel-keitel** [comment]
All suggestions from the original PR: https://github.com/gfx-rs/wgpu/pull/3020#pullrequestreview-1139691697
are applied. All other comments where considered and resulted in code changes.

I will look over the API differences between Vulkan and DX12 Before I consider this PR "Ready for review".

**Comment by daniel-keitel** [comment]
I made some changes to the Api to improve DX12 Compatibility:
- Single AS build instead of multiple at once
- No support for indirect AS build

And added some features:
- AABB support (untested)
- Support for Geometry Flags
- Support for multiple geometries in single BLAS

In regards to the previously requested changes:
- https://github.com/gfx-rs/wgpu/pull/3020#discussion_r993877518: Surrounding code changed
-  https://github.com/gfx-rs/wgpu/pull/3020#discussion_r993877876: Done
- https://github.com/gfx-rs/wgpu/pull/3020#discussion_r993878614: Done
- https://github.com/gfx-rs/wgpu/pull/3020#discussion_r993879926: We can't really hide it since this structure may be used inside compute shaders. I know of no nice way of handling *C bit fields* in rust. I added some helper methods inside the example, which could be added to core or hal. (At least the struct is identical for Vk and Dx12)
- https://github.com/gfx-rs/wgpu/pull/3020#discussion_r993880556: Done

I have skimmed parts of the Metal specification and it seams fairly compatible,
the main issue I found is that buffer contents are different.
E.g.: Vulkan, DirectX12 and Metal all require  the instance data inside a buffer,
the data structure inside is the same for Vk, Dx but different for Metal.

**Comment by JMS55** [comment]
Hi, super excited to use raytracing! Is there any way I can help with this PR?

**Comment by RDambrosio016** [comment]
> Single AS build instead of multiple at once

Wouldn't it be better to still have this, but just implement it as multiple build calls on dx12?

**Comment by daniel-keitel** [comment]
@JMS55, everything is implemented for this PR, only the changelog and some documentation is missing.
I am currently looking into DX12 to implement it next, if you have experience with,  some help would be appreciated, since I have no experience with DX12. 

I am curios, in which additional features beyond this PR are you interested in.

**Comment by RDambrosio016** [comment]
I looked into the DX12 impl a bit and it seems to be really easy with regards to the dx12 code, but a massive pain with regards to the bindings. wgpu-hal has been switching to winapi (and eventually windows-rs), but winapi does not have DXR things because [this pr](https://github.com/retep998/winapi-rs/pull/812) was never merged. I tried reimplementing those bindings in a hacky file inside wgpu-hal, but that is also a pain because it requires a different graphics command list interface (ID3D12GraphicsCommandList4).

**Comment by JMS55** [comment]
@daniel-keitel I'm interested in using ray tracing in [Bevy](https://bevyengine.org) for real time global illumination. So I'm waiting for high level wgpu APIs. In terms of features, compaction seems important for memory savings for static meshes. Beyond that, I'm not familiar enough with RT APIs and performance guidelines to know if there's anything else important missing. I assume we'll have to learn what APIs we need to tweak or expose as we go, once users start experimenting with them.

I don't have much experience with RT or DX12, but I can probably help and learn as I go if there's concrete issues you need help with.

**Comment by RDambrosio016** [comment]
The raytracing pipeline is the other very complex part of raytracing. You basically never want to explicitly sample the BVH from a normal shader. A proper raytracing pipeline with raygen, hit, closesthit, miss, etc shaders is the next big step. That includes a lot of other slightly painful things like SBT records and such,

**Comment by JMS55** [comment]
I did some cursory googling, it looks like ray queries have have similar performance to full ray pipelines? Is this info wrong, or just incomplete?

**Comment by RDambrosio016** [comment]
Yes it shouldn't be a huge difference, but the different shader types have more features, and it allows the GPU to schedule them in a much more custom manner. For example, SER (Shader Execution Reordering) on ada lovelace GPUs. It also hels to separate everything, its much cleaner.

**Comment by daniel-keitel** [comment]
@JMS55 Thanks, I will change the API to allow multiple builds at once. I'm interested in using it inside Bevy too.

**Comment by daniel-keitel** [comment]
> I did some cursory googling, it looks like ray queries have have similar performance to full ray pipelines? Is this info wrong, or just incomplete?

As far I know it is true for NVIDIA Gpus, AMD ones had performance issues the last time I checked.

But I definitely agree, that we need the rt pipeline, for proper rt support.

The question I am asking myself is, should we implement the low level first (Dx12, compaction, ...),
or should we move to wgpu-core and wgpu?

I couldn't find a clear answer how, we would implement it in the wgpu-core and wgpu crates.
I expect we would simply implement it inside the existing creates and hide it behind a feature flag?

**Comment by JMS55** [comment]
I would implement RT pipelines, and then move onto wgpu-core/wgpu. DX12 can be postponed until Vulkan works, so long as you ensure that the APIs are designed to work well for both VK/DX12, and aren't VK specific. Compaction is really important for memory usage, but not functionality, and can be added later. 

Probably not a feature flag. Maybe put the RT stuff in the [util](https://docs.rs/wgpu/latest/wgpu/util/index.html) module? Up to the maintainers.

**Comment by daniel-keitel** [comment]
Everything is implemented for this PR (and somewhat documented).
I will continue by working on the ray-tracing pipeline.

**Comment by daniel-keitel** [comment]
@JMS55 are you interested in working together on the next PR?

**Comment by JMS55** [comment]
> @JMS55 are you interested in working together on the next PR?

I would be happy to :). Unfortunately, I don't have my desktop for another week, so I can't test any ray tracing stuff. I can still help though, as it's just wrapping other APIs. I don't need to test it directly to work on it. Just let me know how you want to divide the work.

File: wgpu-hal/src/lib.rs
Comment (JMS55) [inline wgpu-hal/src/lib.rs:574]:
```suggestion
    /// with the exception of updates in the same descriptor.
```

**Comment by JMS55** [comment]
> @JMS55 it would great if we could talk about the Api surface we want to create.

Sure. Does discord messaging work? I can email my discord ID to the email in your user profile.

**Comment by expenses** [comment]
@daniel-keitel I have access to a GPU with ray tracing support again, so let me know if I can help in any way to push this forward.

**Comment by daniel-keitel** [comment]
@expenses This PR should be "complete" ,general feedback is appreciated, (for #3607 as well).
And It would be great if you would join the discussion about support in wgpu-core, in the matrix thread:
https://matrix.to/#/!FZyQrssSlHEZqrYcOb:matrix.org/$vuQn2Lx30kOuFttuHHTpVZbKcZvCMScAMzTDi8r3s2A?via=matrix.org&via=mozilla.org&via=ralith.com

**Comment by daniel-keitel** [comment]
After thinking about the higher level APIs, I will likely change this API to take, iterators instead of slices, to reduce temporary allocations.

**Comment by daniel-keitel** [comment]
> After thinking about the higher level APIs, I will likely change this API to take, iterators instead of slices, to reduce temporary allocations.

I will leave it as it is for now, since Iterators over Enums containing (boxed dyn) Iterators, seam kind of wrong to me.
It additionally makes debugging more difficult.

**Comment by Jaisiero** [comment]
Hi there!

I downloaded your commit and tried to run the example on Win11 inside this path:
_wgpu\wgpu-hal\examples\ray-traced-triangle_

Gives me this error:
`thread 'main' panicked at 'Selected backend is not supported: InstanceError', wgpu-hal\examples\ray-traced-triangle\main.rs:1080:43`


Am I doing something wrong?

Cheers,
Jaime.

**Comment by daniel-keitel** [comment]
@Jaisiero the error handling, isn't very verbose in the example. 
Can you tell me what gpu you have? (I suspect it is incompatible, I will add more verbose error messages, if that's not the case)

**Comment by Jaisiero** [comment]
> @Jaisiero the error handling, isn't very verbose in the example. Can you tell me what gpu you have? (I suspect it is incompatible, I will add more verbose error messages, if that's not the case)

After debugging a little, it seems like in main.rs:238 adapters are empty. Maybe I am running it in a wrong way?

**Comment by daniel-keitel** [comment]
> > @Jaisiero the error handling, isn't very verbose in the example. Can you tell me what gpu you have? (I suspect it is incompatible, I will add more verbose error messages, if that's not the case)
> 
> After debugging a little, it seems like in main.rs:238 adapters are empty. Maybe I am running it in a wrong way?

@Jaisiero If that's the case and 'halmark' also doesn't work, then the issue is most likely due to wrong features.
Did you start it by executing `cargo run --example ray-traced-triangle`?

**Comment by Jaisiero** [comment]
> @Jaisiero the error handling, isn't very verbose in the example. Can you tell me what gpu you have? (I suspect it is incompatible, I will add more verbose error messages, if that's not the case)

I have a RTX 3090. It definitely supports ray tracing but maybe I am not enabling vulkan in my PC. Do I must configure wgpu to use vulkan backend?

**Comment by Jaisiero** [comment]
> > > @Jaisiero the error handling, isn't very verbose in the example. Can you tell me what gpu you have? (I suspect it is incompatible, I will add more verbose error messages, if that's not the case)
> > 
> > 
> > After debugging a little, it seems like in main.rs:238 adapters are empty. Maybe I am running it in a wrong way?
> 
> @Jaisiero If that's the case and 'halmark' also doesn't work, then the issue is most likely due to wrong features. Did you start it by executing `cargo run --example ray-traced-triangle`?

Allright, I tried again from the main wgpu folder and now the example works (idk what i did previously).

So from here, the problem I see is that WGSL doesn't support .rgen, .rchit, .rmiss and .rahit. Do you think I could use native GLSL spv shaders to build the ray tracing pipeline? Next step should be to build the Shader Binding Table. This is gonna be a lot of work haha.

Cheers,
Jaime.

**Comment by daniel-keitel** [comment]
@Jaisiero I have already implemented the ray-pipeline (with sbt) for wgpu-hal with spv shaders in PR #3607.
My focus currently lies in PR #3631, to add raq-query and acceleration structure support to the wgpu-core and wgpu api,
I will continue working on that next week.

For what do you want to use ray tracing?

**Comment by Jaisiero** [comment]
> @Jaisiero I have already implemented the ray-pipeline (with sbt) for wgpu-hal with spv shaders in PR #3607. My focus currently lies in PR #3631, to add raq-query and acceleration structure support to the wgpu-core and wgpu api, I will continue working on that next week.
> 
> For what do you want to use ray tracing?

I would like to build a voxel engine in wgpu and using hardware accel. struct would speed it up! I will need AABB primitives (procedural) for it. Actually I made a primitive one in raw Vulkan. Wouldn't it be great to have hardware accel. RT on web? It should be the first tech to achieve it, right?

**Comment by expenses** [comment]
Hey @daniel-keitel, this branch is a little behind so I just merged trunk into it: https://github.com/expenses/wgpu/tree/hal-acceleration-structures, merge commit: https://github.com/expenses/wgpu/commit/fec5e40c125a388f2ebb37cf4cc8827c3d10aa57

**Comment by JMS55** [comment]
@expenses merged trunk into this branch again based off of yours: https://github.com/expenses/wgpu/pull/2. Will try and rebase https://github.com/gfx-rs/wgpu/pull/3631 next.

**Comment by cwfitzgerald** [comment]
Could someone tell me how all these MRs are related, I'm having a lot of trouble following them.

**Comment by JMS55** [comment]
This PR is just the wgpu_hal VK backend for RT. No metal, dx12, wpgu-core, or wgpu stuff.

This other PR https://github.com/gfx-rs/wgpu/pull/3631 is building on that and implements RT support at the wgpu-core and wgpu level.

EDIT: BLAS/TLAS support only. No RT pipelines atm. These are meant to be used with inline RT queries, as you picked up on.

**Comment by cwfitzgerald** [comment]
Cool - I've renamed them - is this ready for review and landing?

**Comment by JMS55** [comment]
I believe so? I'll leave it up to @daniel-keitel to confirm, though.

**Comment by daniel-keitel** [comment]
It is ready for review, after I had a look at the failed tests.

File: wgpu-hal/Cargo.toml
Comment (cwfitzgerald) [inline wgpu-hal/Cargo.toml]: Let's sort this alphabetically

File: wgpu-hal/examples/ray-traced-triangle/shader.comp
Comment (cwfitzgerald) [inline wgpu-hal/examples/ray-traced-triangle/shader.comp]: Why are we using glsl for this? Shouldn't wgsl be supported?

File: wgpu-hal/examples/ray-traced-triangle/shader.comp
Comment (cwfitzgerald) [inline wgpu-hal/examples/ray-traced-triangle/shader.comp]: Have some mixed indenting

File: wgpu-hal/src/vulkan/device.rs
Comment (cwfitzgerald) [inline wgpu-hal/src/vulkan/device.rs]: This can just be an as_ref().expect() in this and other places

File: wgpu-hal/src/vulkan/command.rs
Comment (cwfitzgerald) [inline wgpu-hal/src/vulkan/command.rs]: What's up with these?

File: wgpu-hal/examples/ray-traced-triangle/main.rs
Comment (cwfitzgerald) [inline wgpu-hal/examples/ray-traced-triangle/main.rs]: This is now wgt::InstanceFlags::default()

File: wgpu-hal/examples/ray-traced-triangle/main.rs
Comment (cwfitzgerald) [inline wgpu-hal/examples/ray-traced-triangle/main.rs]: We should name Dxc here, as we need it for d3d

File: wgpu-hal/examples/ray-traced-triangle/main.rs
Comment (cwfitzgerald) [inline wgpu-hal/examples/ray-traced-triangle/main.rs:700]: Why isn't this just a normal barrier?

**Review by cwfitzgerald — CHANGES_REQUESTED** [review]
Sorry for taking so long to review this, we'll be better about this going forward.

This looks great! Some smaller nits and needs a rebase, but let's get this in!

**Comment by cwfitzgerald** [comment]
(Please re-request my review when changes are addressed!)

File: wgpu-hal/src/vulkan/command.rs
Comment (daniel-keitel) [reply wgpu-hal/src/vulkan/command.rs]: must have forgotten to remove those

File: wgpu-hal/examples/ray-traced-triangle/main.rs
Comment (daniel-keitel) [reply wgpu-hal/examples/ray-traced-triangle/main.rs:700]: As far as I can tell Vulkan and DX12 require quite different syncronisation, (since dx uses a buffer for the acceleration structure while vulkan has its own resource type).
So this function is in preperation of the DX12 implementation.

File: wgpu-hal/examples/ray-traced-triangle/shader.comp
Comment (daniel-keitel) [reply wgpu-hal/examples/ray-traced-triangle/shader.comp]: It was not yet supported in naga, when I started working on this PR.
I changed It to wgsl.

**Comment by JMS55** [comment]
Looks like this PR needs rebasing btw. Happy to see more progress on it though, thank you!

**Pull request merged**
